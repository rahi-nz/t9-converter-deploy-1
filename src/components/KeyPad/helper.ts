export const PHONE_BUTTONS = [
  {
    disabled: true,
    subtitle: '',
    title: '1',
    value: 1,
  },
  {
    disabled: false,
    subtitle: 'abc',
    title: '2',
    value: 2,
  },
  {
    disabled: false,
    subtitle: 'def',
    title: '3',
    value: 3,
  },
  {
    disabled: false,
    subtitle: 'ghi',
    title: '4',
    value: 4,
  },
  {
    disabled: false,
    subtitle: 'jkl',
    title: '5',
    value: 5,
  },
  {
    disabled: false,
    subtitle: 'mno',
    title: '6',
    value: 6,
  },
  {
    disabled: false,
    subtitle: 'pqrs',
    title: '7',
    value: 7,
  },
  {
    disabled: false,
    subtitle: 'tuv',
    title: '8',
    value: 8,
  },
  {
    disabled: false,
    subtitle: 'wxyz',
    title: '9',
    value: 9,
  },
  {
    disabled: true,
    subtitle: '',
    title: '*',
    value: '*',
  },
  {
    disabled: true,
    subtitle: '',
    title: '0',
    value: 0,
  },
  {
    disabled: true,
    subtitle: '',
    title: '#',
    value: '#',
  },
];
