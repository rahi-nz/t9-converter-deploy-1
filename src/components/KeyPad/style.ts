import styled from 'styled-components';

interface Prop {
  readonly enable: boolean;
}

export const Filter = styled.section`
  width: 200px;
  padding: 20px;
`;

export const Container = styled.section`
  width: 240px;
  background: #fff;
  border-radius: 40px;
  border: 1px solid #48b8a5;
  box-shadow: rgb(0 0 0 / 8%) 0px 1px 12px;
`;

export const Input = styled.input`
  width: 170px;
  border: none;
  padding: 30px;
  border-radius: 40px;
  font-size: 18px;
  margin: 5px;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  :focus-visible {
    outline: none;
  }
`;

export const ListContainer = styled.div`
  min-height: 60px;
`;

export const Word = styled.span.attrs(() => ({
  'data-test-id': 'suggested',
}))`
  background: #e0e0e0;
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 10px 0;
  border-radius: 6px;
`;

export const Row = styled.div`
  padding: 10px;
  display: flex;
  justify-content: space-between;
`;

export const ClearBtn = styled.div<Prop>`
  ${props => {
    if (props.enable)
      return `
          cursor: pointer;
        `;
  }}
`;

export const KeyPadContainer = styled.section`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
`;
