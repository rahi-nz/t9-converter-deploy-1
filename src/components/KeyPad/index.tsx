import { debounce } from 'lodash';
import { FixedSizeList as List } from 'react-window';
import { useCallback, useState } from 'react';
import { server } from '../../config';
import Image from 'next/image';
import {
  Filter,
  Container,
  KeyPadContainer,
  Input,
  Word,
  ListContainer,
  Row,
  ClearBtn,
} from './style';
import { PHONE_BUTTONS } from './helper';
import Button from '../Button';

const GUTTER_SIZE = 5;

const KeyPad = () => {
  const [input, setInput] = useState<string>('');
  const [wordList, setWordList] = useState<string[]>();
  const [onlyRealWords, setOnlyRealWords] = useState<boolean>(false);
  const validInput = input.length > 0;
  /**
   * debounce: to fire it only when the user has finished typing
   * useCallback: store the reference
   */
  const debounceLoadData = useCallback(
    debounce(async (numbers, realWord) => {
      if (numbers.length > 0) {
        const requestBody = {
          onlyRealWords: realWord,
          numbers,
        };
        const response = await fetch(`${server}/api/converter`, {
          method: 'post',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(requestBody),
        }).then(res => res.json());
        setWordList(response);
      } else {
        setWordList([]);
      }
    }, 1000),
    [],
  );

  const onClick = key => {
    setInput(prevState => prevState.concat(key));
    debounceLoadData(input.concat(key), onlyRealWords);
  };

  const onRealWord = () => {
    setOnlyRealWords(prevState => !prevState);
    debounceLoadData(input, !onlyRealWords);
  };

  const onClear = () => {
    if (validInput) {
      setInput(prevState => prevState.slice(0, -1));
      debounceLoadData(input.slice(0, -1), onlyRealWords);
    }
  };

  const Column = ({ index, style }: any) => (
    <Word
      style={{
        ...style,
        left: style.left + GUTTER_SIZE,
        top: style.top + GUTTER_SIZE,
        width: style.width - GUTTER_SIZE,
        height: 25,
      }}
    >
      {wordList[index]}
    </Word>
  );

  return (
    <>
      <Filter>
        <input
          name="realWord"
          id="realWord"
          type="checkbox"
          checked={onlyRealWords}
          onChange={onRealWord}
        />
        <label htmlFor="realWord">Show only real words</label>
      </Filter>
      <Container>
        <Input type="number" defaultValue={input} disabled />
        <ListContainer>
          {wordList && (
            <List
              height={60}
              itemCount={wordList.length}
              itemSize={100}
              layout="horizontal"
              width={240}
            >
              {Column}
            </List>
          )}
        </ListContainer>
        <Row>
          <Image width={30} height={30} src="/icons/plus.svg" alt="kiwi" />
          <ClearBtn enable={validInput}>
            <Image
              width={30}
              height={30}
              src="/icons/clear-icon.svg"
              alt="kiwi"
              onClick={onClear}
            />
          </ClearBtn>
        </Row>
        <KeyPadContainer>
          {PHONE_BUTTONS.map(button => (
            <Button
              handleClick={() => onClick(button.value)}
              key={button.value}
              disabled={button.disabled}
              title={button.title}
              subtitle={button.subtitle}
            />
          ))}
        </KeyPadContainer>
      </Container>
    </>
  );
};

export default KeyPad;
