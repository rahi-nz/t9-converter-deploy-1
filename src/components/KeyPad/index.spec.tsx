/**
 * @jest-environment jsdom
 */

import renderer, { act } from 'react-test-renderer';
import { waitFor } from '@testing-library/react';
import KeyPad from './index';
import fetchMock from 'jest-fetch-mock';

describe('<KeyPad/>', () => {
  let rootElement;
  let testRenderer;
  beforeEach(() => {
    // eslint-disable-next-line prefer-const,react/display-name
    rootElement = () => <KeyPad />;
    // eslint-disable-next-line prefer-const
    testRenderer = renderer.create(rootElement());
    fetchMock.resetMocks();
  });

  it('should call fetchData after 1000ms after typing', async () => {
    fetchMock.mockResponseOnce(JSON.stringify(['a', 'b', 'c']));
    await act(async () => {
      testRenderer.root.findAllByType('button')[1].props.onClick();
      await waitFor(
        () => {
          expect(
            testRenderer.root.findAllByProps({ 'data-test-id': 'suggested' }).length,
          ).toBe(3);
        },
        {
          timeout: 1500,
        },
      );
    });
  });

  it('should not call fetchData before 1000ms after typing', async () => {
    await act(async () => {
      testRenderer.root.findAllByType('button')[1].props.onClick();
      await waitFor(
        () => {
          expect(
            testRenderer.root.findAllByProps({ 'data-test-id': 'suggested' }).length,
          ).toBe(0);
        },
        {
          timeout: 500,
        },
      );
    });
  });

  it('Should render correctly', () => {
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
});
