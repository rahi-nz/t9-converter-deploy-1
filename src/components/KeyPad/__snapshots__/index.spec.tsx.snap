// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<KeyPad/> Should render correctly 1`] = `
Array [
  <section
    className="style__Filter-sc-1pdyef4-0 czjbYD"
  >
    <input
      checked={false}
      id="realWord"
      name="realWord"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      htmlFor="realWord"
    >
      Show only real words
    </label>
  </section>,
  <section
    className="style__Container-sc-1pdyef4-1 jmmXFE"
  >
    <input
      className="style__Input-sc-1pdyef4-2 foovYR"
      defaultValue=""
      disabled={true}
      type="number"
    />
    <div
      className="style__ListContainer-sc-1pdyef4-3 kCrRnV"
    />
    <div
      className="style__Row-sc-1pdyef4-5 gwCNqW"
    >
      <div
        style={
          Object {
            "boxSizing": "border-box",
            "display": "inline-block",
            "margin": 0,
            "maxWidth": "100%",
            "overflow": "hidden",
            "position": "relative",
          }
        }
      >
        <div
          style={
            Object {
              "boxSizing": "border-box",
              "display": "block",
              "maxWidth": "100%",
            }
          }
        >
          <img
            alt=""
            aria-hidden={true}
            role="presentation"
            src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMzAiIGhlaWdodD0iMzAiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
            style={
              Object {
                "border": "none",
                "display": "block",
                "margin": 0,
                "maxWidth": "100%",
                "padding": 0,
              }
            }
          />
        </div>
        <noscript>
          <img
            alt="kiwi"
            decoding="async"
            src="/icons/plus.svg"
            srcSet="/_next/image?url=%2Ficons%2Fplus.svg&w=32&q=75 1x, /_next/image?url=%2Ficons%2Fplus.svg&w=64&q=75 2x"
            style={
              Object {
                "border": "none",
                "bottom": 0,
                "boxSizing": "border-box",
                "display": "block",
                "height": 0,
                "left": 0,
                "margin": "auto",
                "maxHeight": "100%",
                "maxWidth": "100%",
                "minHeight": "100%",
                "minWidth": "100%",
                "objectFit": undefined,
                "objectPosition": undefined,
                "padding": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "width": 0,
              }
            }
          />
        </noscript>
        <img
          alt="kiwi"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style={
            Object {
              "border": "none",
              "bottom": 0,
              "boxSizing": "border-box",
              "display": "block",
              "height": 0,
              "left": 0,
              "margin": "auto",
              "maxHeight": "100%",
              "maxWidth": "100%",
              "minHeight": "100%",
              "minWidth": "100%",
              "objectFit": undefined,
              "objectPosition": undefined,
              "padding": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "width": 0,
            }
          }
        />
      </div>
      <div
        className="style__ClearBtn-sc-1pdyef4-6"
      >
        <div
          style={
            Object {
              "boxSizing": "border-box",
              "display": "inline-block",
              "margin": 0,
              "maxWidth": "100%",
              "overflow": "hidden",
              "position": "relative",
            }
          }
        >
          <div
            style={
              Object {
                "boxSizing": "border-box",
                "display": "block",
                "maxWidth": "100%",
              }
            }
          >
            <img
              alt=""
              aria-hidden={true}
              role="presentation"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMzAiIGhlaWdodD0iMzAiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
              style={
                Object {
                  "border": "none",
                  "display": "block",
                  "margin": 0,
                  "maxWidth": "100%",
                  "padding": 0,
                }
              }
            />
          </div>
          <noscript>
            <img
              alt="kiwi"
              decoding="async"
              onClick={[Function]}
              src="/icons/clear-icon.svg"
              srcSet="/_next/image?url=%2Ficons%2Fclear-icon.svg&w=32&q=75 1x, /_next/image?url=%2Ficons%2Fclear-icon.svg&w=64&q=75 2x"
              style={
                Object {
                  "border": "none",
                  "bottom": 0,
                  "boxSizing": "border-box",
                  "display": "block",
                  "height": 0,
                  "left": 0,
                  "margin": "auto",
                  "maxHeight": "100%",
                  "maxWidth": "100%",
                  "minHeight": "100%",
                  "minWidth": "100%",
                  "objectFit": undefined,
                  "objectPosition": undefined,
                  "padding": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "width": 0,
                }
              }
            />
          </noscript>
          <img
            alt="kiwi"
            decoding="async"
            onClick={[Function]}
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style={
              Object {
                "border": "none",
                "bottom": 0,
                "boxSizing": "border-box",
                "display": "block",
                "height": 0,
                "left": 0,
                "margin": "auto",
                "maxHeight": "100%",
                "maxWidth": "100%",
                "minHeight": "100%",
                "minWidth": "100%",
                "objectFit": undefined,
                "objectPosition": undefined,
                "padding": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "width": 0,
              }
            }
          />
        </div>
      </div>
    </div>
    <section
      className="style__KeyPadContainer-sc-1pdyef4-7 kkNFth"
    >
      <button
        className="style__DialBtn-niwgt9-0 hWtKCf"
        disabled={true}
        onClick={[Function]}
      >
        <div
          className="style__Title-niwgt9-1 iIrczW"
        >
          1
        </div>
        <div
          className="style__Subtitle-niwgt9-2 efnWUX"
        >
          
        </div>
      </button>
      <button
        className="style__DialBtn-niwgt9-0 dtWhVz"
        disabled={false}
        onClick={[Function]}
      >
        <div
          className="style__Title-niwgt9-1 iIrczW"
        >
          2
        </div>
        <div
          className="style__Subtitle-niwgt9-2 efnWUX"
        >
          abc
        </div>
      </button>
      <button
        className="style__DialBtn-niwgt9-0 dtWhVz"
        disabled={false}
        onClick={[Function]}
      >
        <div
          className="style__Title-niwgt9-1 iIrczW"
        >
          3
        </div>
        <div
          className="style__Subtitle-niwgt9-2 efnWUX"
        >
          def
        </div>
      </button>
      <button
        className="style__DialBtn-niwgt9-0 dtWhVz"
        disabled={false}
        onClick={[Function]}
      >
        <div
          className="style__Title-niwgt9-1 iIrczW"
        >
          4
        </div>
        <div
          className="style__Subtitle-niwgt9-2 efnWUX"
        >
          ghi
        </div>
      </button>
      <button
        className="style__DialBtn-niwgt9-0 dtWhVz"
        disabled={false}
        onClick={[Function]}
      >
        <div
          className="style__Title-niwgt9-1 iIrczW"
        >
          5
        </div>
        <div
          className="style__Subtitle-niwgt9-2 efnWUX"
        >
          jkl
        </div>
      </button>
      <button
        className="style__DialBtn-niwgt9-0 dtWhVz"
        disabled={false}
        onClick={[Function]}
      >
        <div
          className="style__Title-niwgt9-1 iIrczW"
        >
          6
        </div>
        <div
          className="style__Subtitle-niwgt9-2 efnWUX"
        >
          mno
        </div>
      </button>
      <button
        className="style__DialBtn-niwgt9-0 dtWhVz"
        disabled={false}
        onClick={[Function]}
      >
        <div
          className="style__Title-niwgt9-1 iIrczW"
        >
          7
        </div>
        <div
          className="style__Subtitle-niwgt9-2 efnWUX"
        >
          pqrs
        </div>
      </button>
      <button
        className="style__DialBtn-niwgt9-0 dtWhVz"
        disabled={false}
        onClick={[Function]}
      >
        <div
          className="style__Title-niwgt9-1 iIrczW"
        >
          8
        </div>
        <div
          className="style__Subtitle-niwgt9-2 efnWUX"
        >
          tuv
        </div>
      </button>
      <button
        className="style__DialBtn-niwgt9-0 dtWhVz"
        disabled={false}
        onClick={[Function]}
      >
        <div
          className="style__Title-niwgt9-1 iIrczW"
        >
          9
        </div>
        <div
          className="style__Subtitle-niwgt9-2 efnWUX"
        >
          wxyz
        </div>
      </button>
      <button
        className="style__DialBtn-niwgt9-0 hWtKCf"
        disabled={true}
        onClick={[Function]}
      >
        <div
          className="style__Title-niwgt9-1 iIrczW"
        >
          *
        </div>
        <div
          className="style__Subtitle-niwgt9-2 efnWUX"
        >
          
        </div>
      </button>
      <button
        className="style__DialBtn-niwgt9-0 hWtKCf"
        disabled={true}
        onClick={[Function]}
      >
        <div
          className="style__Title-niwgt9-1 iIrczW"
        >
          0
        </div>
        <div
          className="style__Subtitle-niwgt9-2 efnWUX"
        >
          
        </div>
      </button>
      <button
        className="style__DialBtn-niwgt9-0 hWtKCf"
        disabled={true}
        onClick={[Function]}
      >
        <div
          className="style__Title-niwgt9-1 iIrczW"
        >
          #
        </div>
        <div
          className="style__Subtitle-niwgt9-2 efnWUX"
        >
          
        </div>
      </button>
    </section>
  </section>,
]
`;
